File: packages/skill/src/format.ts
/**
 * Render the `<skill>` XML wrap that carries an activated skill body into
 * conversation. The wrapper is load-bearing for auto-compaction: harness
 * summarizers can recognize the marker and preserve the full body.
 */
export interface FormatSkillArgs {
  readonly name: string;
  readonly dir: string;
  readonly frontmatter: Readonly<Record<string, unknown>>;
  readonly body: string;
  readonly resources: readonly string[];
  readonly bytes: number;
}

export function formatSkill(args: FormatSkillArgs): string {
  const fmSerialized = serializeFrontmatter(args.frontmatter);
  const instructions = args.body;
  const resourcesBlock =
    args.resources.length === 0
      ? ""
      : `<resources>\n${args.resources.join("\n")}\n</resources>\n`;
  const scriptsDir = `${args.dir}/scripts`;
  const referencesDir = `${args.dir}/references`;
  const hint = `(Skill "${args.name}" activated. Body is ${args.bytes} bytes. Scripts available via bash(${scriptsDir}/<name>). References via read(${referencesDir}/<name>).)`;

  return [
    `<skill name="${args.name}" dir="${args.dir}">`,
    `<frontmatter>`,
    fmSerialized,
    `</frontmatter>`,
    `<instructions>`,
    instructions,
    `</instructions>`,
    resourcesBlock.trimEnd(),
    `</skill>`,
    hint,
  ]
    .filter((s) => s.length > 0)
    .join("\n");
}

export function formatAlreadyLoaded(name: string): string {
  return `(Skill "${name}" is already active in this session. No new content was added.)`;
}

export interface FormatNotFoundArgs {
  readonly name: string;
  readonly siblings: readonly string[];
}

export function formatNotFound(args: FormatNotFoundArgs): string {
  if (args.siblings.length === 0) {
    return `(No skill matches "${args.name}". Check the catalog for installed skill names.)`;
  }
  return `(No skill matches "${args.name}". Did you mean: ${args.siblings.join(", ")}? Run with a listed name from the catalog.)`;
}

/**
 * Re-serialize the parsed frontmatter back to YAML-ish. We don't need
 * round-trip fidelity — the point is so the model sees the skill's
 * declared metadata alongside the body. Compact form, single-line per
 * key where possible.
 */
function serializeFrontmatter(
  fm: Readonly<Record<string, unknown>>,
): string {
  const lines: string[] = [];
  for (const [key, value] of Object.entries(fm)) {
    lines.push(renderKeyValue(key, value, 0));
  }
  return lines.join("\n");
}

function renderKeyValue(key: string, value: unknown, indent: number): string {
  const pad = "  ".repeat(indent);
  if (value === null || value === undefined) return `${pad}${key}:`;
  if (typeof value === "boolean" || typeof value === "number") {
    return `${pad}${key}: ${String(value)}`;
  }
  if (typeof value === "string") {
    if (value.includes("\n")) {
      const body = value
        .split("\n")
        .map((l) => `${pad}  ${l}`)
        .join("\n");
      return `${pad}${key}: |\n${body}`;
    }
    return `${pad}${key}: ${quoteIfNeeded(value)}`;
  }
  if (Array.isArray(value)) {
    if (value.length === 0) return `${pad}${key}: []`;
    const items = value.map((v) => quoteIfNeeded(String(v))).join(", ");
    return `${pad}${key}: [${items}]`;
  }
  if (typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>);
    if (entries.length === 0) return `${pad}${key}: {}`;
    const nested = entries
      .map(([k, v]) => renderKeyValue(k, v, indent + 1))
      .join("\n");
    return `${pad}${key}:\n${nested}`;
  }
  return `${pad}${key}: ${String(value)}`;
}

function quoteIfNeeded(s: string): string {
  if (/^[A-Za-z0-9_\-\.\/]+$/.test(s)) return s;
  return JSON.stringify(s);
}

File: crates/skill/src/format.rs
use serde_json::{Map, Value};

pub struct FormatSkillArgs<'a> {
    pub name: &'a str,
    pub dir: &'a str,
    pub frontmatter: &'a Value,
    pub body: &'a str,
    pub resources: &'a [String],
    pub bytes: u64,
}

pub fn format_skill(args: FormatSkillArgs<'_>) -> String {
    let fm_serialized = serialize_frontmatter(args.frontmatter);
    let resources_block = if args.resources.is_empty() {
        String::new()
    } else {
        format!("<resources>\n{}\n</resources>", args.resources.join("\n"))
    };
    let hint = format!(
        "(Skill \"{}\" activated. Body is {} bytes. Scripts available via bash({}/scripts/<name>). References via read({}/references/<name>).)",
        args.name, args.bytes, args.dir, args.dir
    );

    let mut parts: Vec<String> = vec![
        format!("<skill name=\"{}\" dir=\"{}\">", args.name, args.dir),
        "<frontmatter>".to_string(),
        fm_serialized,
        "</frontmatter>".to_string(),
        "<instructions>".to_string(),
        args.body.to_string(),
        "</instructions>".to_string(),
    ];
    if !resources_block.is_empty() {
        parts.push(resources_block);
    }
    parts.push("</skill>".to_string());
    parts.push(hint);
    parts.into_iter().filter(|s| !s.is_empty()).collect::<Vec<_>>().join("\n")
}

pub fn format_already_loaded(name: &str) -> String {
    format!(
        "(Skill \"{}\" is already active in this session. No new content was added.)",
        name
    )
}

pub fn format_not_found(name: &str, siblings: &[String]) -> String {
    if siblings.is_empty() {
        format!(
            "(No skill matches \"{}\". Check the catalog for installed skill names.)",
            name
        )
    } else {
        format!(
            "(No skill matches \"{}\". Did you mean: {}? Run with a listed name from the catalog.)",
            name,
            siblings.join(", ")
        )
    }
}

/// Re-serialize frontmatter to YAML-ish form. Not round-trip fidelity —
/// the model just needs to see the declared metadata above the body.
fn serialize_frontmatter(fm: &Value) -> String {
    let mut lines: Vec<String> = Vec::new();
    if let Value::Object(map) = fm {
        for (k, v) in map {
            lines.push(render_kv(k, v, 0));
        }
    }
    lines.join("\n")
}

fn render_kv(key: &str, value: &Value, indent: usize) -> String {
    let pad = "  ".repeat(indent);
    match value {
        Value::Null => format!("{}{}:", pad, key),
        Value::Bool(b) => format!("{}{}: {}", pad, key, b),
        Value::Number(n) => format!("{}{}: {}", pad, key, n),
        Value::String(s) => {
            if s.contains('\n') {
                let body = s
                    .split('\n')
                    .map(|line| format!("{}  {}", pad, line))
                    .collect::<Vec<_>>()
                    .join("\n");
                format!("{}{}: |\n{}", pad, key, body)
            } else {
                format!("{}{}: {}", pad, key, quote_if_needed(s))
            }
        }
        Value::Array(arr) => {
            if arr.is_empty() {
                format!("{}{}: []", pad, key)
            } else {
                let items: Vec<String> = arr
                    .iter()
                    .map(|v| match v {
                        Value::String(s) => quote_if_needed(s),
                        other => serde_json::to_string(other).unwrap_or_default(),
                    })
                    .collect();
                format!("{}{}: [{}]", pad, key, items.join(", "))
            }
        }
        Value::Object(map) => render_map(key, map, indent),
    }
}

fn render_map(key: &str, map: &Map<String, Value>, indent: usize) -> String {
    let pad = "  ".repeat(indent);
    if map.is_empty() {
        return format!("{}{}: {{}}", pad, key);
    }
    let nested: Vec<String> = map.iter().map(|(k, v)| render_kv(k, v, indent + 1)).collect();
    format!("{}{}:\n{}", pad, key, nested.join("\n"))
}

fn quote_if_needed(s: &str) -> String {
    let simple = !s.is_empty()
        && s.chars().all(|c| {
            c.is_ascii_alphanumeric() || c == '_' || c == '-' || c == '.' || c == '/'
        });
    if simple {
        s.to_string()
    } else {
        // JSON-style double-quoted escaping handles all edge cases YAML
        // readers tolerate as quoted strings.
        serde_json::to_string(s).unwrap_or_else(|_| format!("\"{}\"", s))
    }
}
